kafka-producer/src/main/java/com/expedia/www/haystack/pipes/kafka/producer/KafkaToKafkaPipeline.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2018 Expedia, Inc.
+ *
+ *       Licensed under the Apache License, Version 2.0 (the "License");
+ *       you may not use this file except in compliance with the License.
+ *       You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *       Unless required by applicable law or agreed to in writing, software
+ *       distributed under the License is distributed on an "AS IS" BASIS,
+ *       WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *       See the License for the specific language governing permissions and
+ *       limitations under the License.
+ *
+ */
+package com.expedia.www.haystack.pipes.kafka.producer;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import com.expedia.open.tracing.Span;
+import com.expedia.www.haystack.pipes.commons.kafka.TagFlattener;
+import com.expedia.www.haystack.pipes.key.extractor.SpanKeyExtractor;
+import com.netflix.servo.util.VisibleForTesting;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.pool2.ObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.streams.kstream.ForeachAction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class KafkaToKafkaPipeline implements ForeachAction<String, Span> {
+    @VisibleForTesting
+    static final String ERROR_MSG =
+            "Exception posting JSON [%s] to Kafka; received message [%s]";
+    static ObjectPool<KafkaCallback> OBJECT_POOL = new GenericObjectPool<>(
+            new CallbackFactory());
+    @VisibleForTesting
+    static Logger logger = LoggerFactory.getLogger(KafkaToKafkaPipeline.class);
+    @VisibleForTesting
+    static Factory factory = new KafkaToKafkaPipeline.Factory();
+    @VisibleForTesting
+    static Counter kafkaProducerCounter;
+    static Counter requestCounter;
+    static Timer kafkaProducerTimer;
+    private final TagFlattener tagFlattener = new TagFlattener();
+    private final Map<SpanKeyExtractor, List<KafkaProducer<String, String>>> kafkaProducerMap;
+
+    public KafkaToKafkaPipeline(MetricRegistry metricRegistry,
+                                Map<SpanKeyExtractor, List<KafkaProducer<String, String>>> kafkaProducerMap) {
+        this.kafkaProducerMap = kafkaProducerMap;
+        requestCounter = metricRegistry.counter("REQUEST");
+        kafkaProducerTimer = metricRegistry.timer("KAFKA_PRODUCER_POST_TIMER");
+        kafkaProducerCounter = metricRegistry.counter("KAFKA_PRODUCER_POST_COUNTER");
+    }
+
+    @Override
+    public void apply(String key, Span value) {
```

Comment:
where is this key used? And rename param value

Reply:
this is a design defined in ForeachAction<String, Span> interface, used by each module. changing this would require a change in every module.

Reply:
ahh, missed it